lib/charms/alertmanager_k8s/v0/alertmanager_remote_configuration.py

```diff
@@ -0,0 +1,459 @@
+        config_file: Path,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """The RemoteConfigurationProvider object factory.
+
+        This factory provides an alternative way of instantiating the RemoteConfigurationProvider.
+        While the default constructor requires passing a config dict, the factory allows using
+        a configuration file path.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            config_file: Path to the Alertmanager configuration file.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+
+        Returns:
+            RemoteConfigurationProvider object
+        """
+        return cls(charm, cls.load_config_file(config_file), relation_name)
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Event handler for RelationJoinedEvent.
+
+        Takes care of pushing Alertmanager configuration to the relation data bag.
+
+        Args:
+            event: Juju event
+        """
+        if not self._charm.unit.is_leader():
+            return
+        if self.alertmanager_config:
+            self.update_relation_data_bag(self.alertmanager_config, event.relation)
+        else:
+            logger.warning("Alertmanager configuration not available. Ignoring...")
+
+    @staticmethod
+    def load_config_file(path: Path) -> dict:
+        """Reads given Alertmanager configuration file and turns it into a dictionary.
+
+        Args:
+            path: Path to the Alertmanager configuration file
+
+        Returns:
+            dict: Alertmanager configuration file in a form of a dictionary
+
+        Raises:
+            ConfigReadError: if a problem with reading given config file happens
+        """
+        try:
+            with open(path, "r") as config_yaml:
+                config = yaml.safe_load(config_yaml)
+            return config
+        except (FileNotFoundError, OSError, yaml.YAMLError) as e:
+            raise ConfigReadError(path) from e
+
+    def update_relation_data_bag(
+        self, alertmanager_config: dict, relation: Optional[Relation]
+    ) -> None:
+        """Updates relation data bag with Alertmanager config and templates.
+
+        Before updating relation data bag, basic sanity check of given configuration is done.
+
+        Args:
+            alertmanager_config: Alertmanager configuration dictionary.
+            relation: Juju Relation object
+        """
+        config = alertmanager_config
+        templates = self._get_templates(config)
+        if config_main_keys_are_valid(config):
+            relation.data[self._charm.app]["alertmanager_config"] = json.dumps(config)  # type: ignore[union-attr]  # noqa: E501
+            relation.data[self._charm.app]["alertmanager_templates"] = json.dumps(templates)  # type: ignore[union-attr]  # noqa: E501
```

Comment:
Iiuc, since this is the provider side then we can have several alertmanagers related here.

1. Seems like there is a loop missing over all relations, for the case that relation is indeed None.
2. Odd that `tox -e static-lib` didn't catch that.

Reply:
Shouldn't this be handled on the provider charm side?

Reply:
Loop added.